test/metabase/query_processor_test/date_time_zone_functions_test.clj

```diff
@@ -319,3 +319,206 @@
+                                 "2022-11-02T00:00:00+00:00"))))  ; 2022-11-01T23:00:00-01:00
+          (mt/with-temporary-setting-values [driver/report-timezone "UTC"]
+            (is (partial= {:hour 744 :day 31 :month 1 :year 0}
+                          (diffs "2022-10-02T01:00:00+01:00"      ; 2022-10-02T00:00:00+00:00
+                                 "2022-11-02T00:00:00+00:00"))))) ; 2022-11-02T00:00:00+00:00
+        (testing "year"
+          (mt/with-temporary-setting-values [driver/report-timezone "Atlantic/Cape_Verde"]
+            (is (partial= {:day 365, :week 52, :month 12, :year 1}
+                          (diffs "2022-10-02T01:00:00+01:00"     ; 2022-10-01T23:00:00-01:00
+                                 "2023-10-02T00:00:00+00:00")))) ; 2023-10-01T23:00:00-01:00
+          (mt/with-temporary-setting-values [driver/report-timezone "UTC"]
+           (is (partial= {:day 365, :week 52, :month 12, :year 1}
+                         (diffs "2022-10-02T01:00:00+01:00"      ; 2022-10-02T00:00:00+00:00
+                                "2023-10-02T00:00:00+00:00"))))) ; 2023-10-02T00:00:00+00:00
+        (testing "hour under a year"
+          (mt/with-temporary-setting-values [driver/report-timezone "Atlantic/Cape_Verde"]
+            (is (partial= {:day 365 :month 12 :year 1}
+                          (diffs "2022-10-02T00:00:00+00:00"     ; 2022-10-01T23:00:00-01:00
+                                 "2023-10-02T00:00:00+01:00")))) ; 2023-10-01T22:00:00-01:00
+          (mt/with-temporary-setting-values [driver/report-timezone "UTC"]
+            (is (partial= {:day 364 :month 11 :year 0}
+                          (diffs "2022-10-02T00:00:00+00:00"          ; 2022-10-02T00:00:00+00:00
+                                 "2023-10-02T00:00:00+01:00"))))))))) ; 2023-10-01T23:00:00+00:00
+
+(deftest datetime-diff-expressions-test
+  (mt/test-drivers (mt/normal-drivers-with-feature :datetime-diff)
+    (mt/dataset sample-dataset
+      (testing "Args can be expressions that return datetime values"
+        (let [diffs (fn [x y]
+                      (let [units [:second :minute :hour :day :week :month :year]]
+                        (->> (mt/run-mbql-query orders
+                               {:limit 1
+                                :expressions (into {} (for [unit units]
+                                                        [(name unit) [:datetime-diff x y unit]]))
+                                :fields (into [] (for [unit units]
+                                                   [:expression (name unit)]))})
+                             mt/rows first
+                             (zipmap units))))]
+          (is (= {:second 31795200, :minute 529920, :hour 8832, :day 368, :week 52, :month 12, :year 1}
+                 (diffs [:datetime-add #t "2022-10-03T00:00:00" 1 "day"] [:datetime-add #t "2023-10-03T00:00:00" 4 "day"])))))
```

Comment:
Instead of running all the queries, sticking the results in a map, and doing a single `(is (= ...))`, I'd suggest taking that map and using `doseq` and doing ones assertion for each. e.g. something like

```clj
(doseq [[unit expected] {:second 31795200, :minute 529920, :hour 8832, :day 368, :week 52, :month 12, :year 1}]
  (testing unit
    (is (= expected
           (mt/run-mbql-query ...)))))
```

it's easier to understand the test output that way especially if one of the queries fails with an Exception or something

Reply:
Doing it that way would run a query per unit, which slows down test runs significantly for bigquery. The tradeoff was worth it for development productivity IMO, which matters more than the gain in readability for errors in CI.